pylabrobot/resources/plate_adapter.py

```diff
@@ -0,0 +1,183 @@
+    the precise anchor location `to` which the plate is moved has to be calculated on the fly.
+
+    This PlateAdapter class is capable of doing so, and is therefore the base class for
+    diverse resources, e.g.:
+    - plate adapters (e.g. for semi- and non-skirted plates which cannot be used on standard
+      carrier sites)
+    - plate adapters for shakers/heater-shakers/temperature control machines
+    - magnetic racks
+    - On-Deck ThermoCyclers (OTDCs)
+
+   Args:
+        name (str): The name of the PlateAdapter.
+        size_x (float): The size of the PlateAdapter in the x dimension.
+        size_y (float): The size of the PlateAdapter in the y dimension.
+        size_z (float): The size of the PlateAdapter in the z dimension.
+        dx (float): The x-coordinate offset for well positioning.
+        dy (float): The y-coordinate offset for well positioning.
+        dz (float): The z-coordinate offset for well positioning, i.e. the outside-bottom
+          of a well.
+        adapter_item_dx (Literal[9.0, 4.5, 2.25], optional): The x-dimension spacing of
+          wells. Defaults to 9.0.
+        adapter_item_dy (Literal[9.0, 4.5, 2.25], optional): The y-dimension spacing of
+        wells. Defaults to 9.0.
+        site_pedestal_z (Optional[float], optional): The z-coordinate of the site pedestal.
+          Defaults to None.
+        category (Optional[str], optional): The category of the PlateAdapter.
+          Defaults to "plate_adapter".
+        model (Optional[str], optional): The model of the PlateAdapter. Defaults to None.
+
+  Examples:
+    1. Using a "magnetic rack" as a PlateAdapter:
+
+      Define a `MFXCarrier`, define a `MFXModule` for deep well plates,
+      assign the DWP module to the MFXCarrier:
+      >>> mfx_carrier_1 = MFX_CAR_L5_base(name='mfx_carrier_1')
+      >>> mfx_carrier_1[1] = mfx_dwp_module_1 = MFX_DWP_module(name="mfx_dwp_module_1")
+
+      Define Alpaqua magnet as a PlateAdapter & assign to mfx_dwp_module_1:
+      >>> Alpaqua_magnum_flx_1 = Alpaqua_magnum_flx(name='Alpaqua_magnum_flx_1')
+      >>> mfx_dwp_module_1.assign_child_resource(Alpaqua_magnum_flx_1)
+
+      Define Cos_96_Rd plate & assign to PlateAdapter:
+      >>> Cos96_plate_1 = Cos_96_Rd(name='Cos96_plate_1')
+      >>> Alpaqua_magnum_flx_1.assign_child_resource(Cos96_plate_1)
+  """
+
+  def __init__(
+    self,
+    name: str,
+    size_x: float, size_y: float, size_z: float,
+    dx: float, dy: float, dz: float,
+    adapter_hole_size_x: float,
+    adapter_hole_size_y: float,
+    site_pedestal_z: float,
+    adapter_hole_dx: float = 9.0,
+    adapter_hole_dy: float = 9.0,
+    category: Optional[str] = None,
+    model: Optional[str] = None
+    ):
+    if adapter_hole_dx not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dx must be one of 9.0, 4.5, or 2.25")
+    if adapter_hole_dy not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dy must be one of 9.0, 4.5, or 2.25")
```

Comment:
does this make things easier?

Reply:
No, I thought it makes PlateAdapter generation safer but I'm happy to remove it to make the class simpler?

Reply:
It seems to me the constraint is artificial, adding a tiny bit of complexity without adding much value. Let's remove it if you agree